_code/InSituTree/R/supervisedSubcluster.R

```diff
@@ -0,0 +1,70 @@
+#' Supervised subclustering using a gene subset and insitutypeML
+#'
+#' @param reference_profiles Matrix of expression profiles, genes x cell types
+#' @param x Counts matrix, cells x genes
+#' @param neg Vector of mean negative controls for each cell
+#' @param quantile_absolute_expression_difference Quantile cutoff for absolute expression.  Default = 0.5. Both quantile cutoffs must be passed to retain gene.
+#' @param quantile_percent_expression_difference Quantile cutoff for percent expression.  Default = 0.5. Both quantile cutoffs must be passed to retain gene.
+#' @param excluded_genes Genes to be excluded during fitting with InSituType
+#' @param cohort Vector of cells' cohort membership
+#'
+#' @import InSituType
+#'
+#' @return Insitutype result list.
+#' @export
+
+supervisedSubcluster <- function(reference_profiles,
+                                  x,
+                                  neg,
+                                  quantile_absolute_expression_difference = 0.5,
+                                  quantile_percent_expression_difference = 0.5,
+                                  excluded_genes = NULL,
+                                  cohort = NULL
+){
+  nsprofiles <- reference_profiles[intersect(colnames(x), rownames(reference_profiles)),]
+  nsprofiles <- apply(nsprofiles, 2, function(x) x/sum(x)*100) # normalize the profiles
+
+  ## Determine which genes have significant differences
+  # Absolute difference
+  diffs <- sapply(rownames(nsprofiles), function(gene){
+    diff(range(nsprofiles[gene, ]))
+  } )
+
+  # Percentage difference
+  percs <- sapply(rownames(nsprofiles), function(gene){
+    diff(range(nsprofiles[gene, ])) / max(nsprofiles[gene,])
+  } )
+
+  # Isolate genes with high absolute and percentage differences
+  use_genes <- names(diffs)[diffs > quantile(diffs, quantile_absolute_expression_difference, na.rm = T) &
+                             percs > quantile(percs, quantile_percent_expression_difference, na.rm = T)]
+
+  # Remove genes that should not be considered.
+  use_genes <- use_genes[!use_genes %in% excluded_genes]
+  message("These are the selected ", length(use_genes),  " genes for subclustering:")
+  dput(use_genes)
+
+  # Warn if number of genes is low
+  if(length(use_genes) < 300){
+    warning(paste0("There are fewer than 300 genes being used to subcluster cells into ",
+                   paste(colnames(reference_profiles), collapse = ", "),
+                   ". These annotations may be unstable."))
+  }
+
+  # verify that all cells have at least 1 count with gene subset
+  nonZeroCount_idx <- Matrix::rowSums(x[,use_genes]) >= 1
+  if(!all(nonZeroCount_idx)) warning(paste0(sum(!nonZeroCount_idx), " cells have 0 counts in the subclustering gene panel."))
```

Comment:
How do you handle the zero-count cells? Are they excluded from all the downstream subclustering? It's worth mentioning in the man that cells with zero counts would not be reported in the return results. Or perhaps, include a vector listing the cell_ID for zero-count cells?

Reply:
I've added the zero count calls back into the summary data.frame with NA values. There's also now a description in the vignette, "Note that any cells which did not have any counts in the genes used for
subclustering will not get a cell type assignment; if this happens in the first
round of annotation, these cells will show 'NA' for all annotation levels. If
this happens at later rounds of annotation, the last assignment will be cascaded
right."

Reply:
Great, but would be good to also add that description to `@details` section of `runInSituTree()`'s man as well.

Reply:
Added the following to the `@details`: 
Any cells which did not have any counts in the genes used for
subclustering will not get a cell type assignment; if this happens in the
first round of annotation, these cells will show 'NA' for all annotation
levels. If this happens at later rounds of annotation, the last assignment
will be cascaded right.
Warnings will be shown if the number of genes chosen is small or if the
counts per cell are low, as both can reduce confidence in the cell type
assignments.